cpp/src/arrow/builder.h

```diff
@@ -129,14 +130,15 @@ class ARROW_EXPORT ArrayBuilder {
 
   // Unsafe operations (don't check capacity/don't resize)
 
-  // Append to null bitmap.
+  // Append to null bitmap, update the length and set finished state to false
   void UnsafeAppendToBitmap(bool is_valid) {
     if (is_valid) {
       BitUtil::SetBit(null_bitmap_data_, length_);
     } else {
       ++null_count_;
     }
     ++length_;
+    is_finished_ = false;
```

Comment:
Seems like this should only be set when the builder is initialized

Reply:
This was my first try. Then I hit the problem that sometimes the builders are initialized but never appended to and never finished, i.e. in DictionaryBuilder. So I changed the logic to "set to false on every append". 

I'll see if I can change it back after the rework I've mentioned before.